src/Symfony/Component/Finder/Gitignore.php

```diff
@@ -14,120 +14,78 @@
+            if ('' !== $line) {
+                if ($isNegative) {
+                    $res = '(?!'.self::lineToRegex($line).')'.$res;
+                } else {
+                    $res = '(?:'.$res.'|'.self::lineToRegex($line).')';
+                }
             }
-
-            $patterns[] = $pattern;
         }
 
-        return sprintf('/^((%s))$/', implode(')|(', $patterns));
+        return '~^(?:'.$res.')~s';
     }
 
-    private static function getRegexFromGitignore(string $gitignorePattern, bool $negative = false): string
+    private static function lineToRegex(string $gitignoreLine): string
     {
-        $regex = '';
-        $isRelativePath = false;
-        // If there is a separator at the beginning or middle (or both) of the pattern, then the pattern is relative to the directory level of the particular .gitignore file itself
-        $slashPosition = strpos($gitignorePattern, '/');
-        if (false !== $slashPosition && \strlen($gitignorePattern) - 1 !== $slashPosition) {
-            if (0 === $slashPosition) {
-                $gitignorePattern = substr($gitignorePattern, 1);
-            }
-
-            $isRelativePath = true;
-            $regex .= '^';
-        }
-
-        if ('/' === $gitignorePattern[\strlen($gitignorePattern) - 1]) {
-            $gitignorePattern = substr($gitignorePattern, 0, -1);
+        if ('' === $gitignoreLine) {
+            return '$f'; // always false
         }
 
-        $iMax = \strlen($gitignorePattern);
-        for ($i = 0; $i < $iMax; ++$i) {
-            $tripleChars = substr($gitignorePattern, $i, 3);
-            if ('**/' === $tripleChars || '/**' === $tripleChars) {
-                $regex .= '.*';
-                $i += 2;
-                continue;
-            }
-
-            $doubleChars = substr($gitignorePattern, $i, 2);
-            if ('**' === $doubleChars) {
-                $regex .= '.*';
-                ++$i;
-                continue;
-            }
-            if ('*/' === $doubleChars) {
-                $regex .= '[^\/]*\/?[^\/]*';
-                ++$i;
-                continue;
-            }
-
-            $c = $gitignorePattern[$i];
-            switch ($c) {
-                case '*':
-                    $regex .= $isRelativePath ? '[^\/]*' : '[^\/]*\/?[^\/]*';
-                    break;
-                case '/':
-                case '.':
-                case ':':
-                case '(':
-                case ')':
-                case '{':
-                case '}':
-                    $regex .= '\\'.$c;
-                    break;
-                default:
-                    $regex .= $c;
+        $slashPos = strpos($gitignoreLine, '/');
+        if (false !== $slashPos && \strlen($gitignoreLine) - 1 !== $slashPos) {
+            if (0 === $slashPos) {
+                $gitignoreLine = substr($gitignoreLine, 1);
             }
+            $isAbsolute = true;
+        } else {
+            $isAbsolute = false;
         }
```

Comment:
What if thr string contains several occurences of `/`? like `/foo/` or `./bar/baz`?
Shouldn't it be? 
```suggestion
        if ($gitignoreLine[0] === '/') {
            $gitignoreLine = substr($gitignoreLine, 1);
            $isAbsolute = true;
        } else {
            $isAbsolute = false;
        }
```

Reply:
No, by spec, citing:

> If there is a separator at the beginning or middle (or both) of the pattern, then the pattern is relative to the directory level of the particular .gitignore file itself. Otherwise the pattern may also match at any level below the .gitignore level.